cocos2d/Platforms/Mac/CCViewMacGL.m

```diff
@@ -0,0 +1,296 @@
+    // When resizing the view, -reshape is called automatically on the main thread
+    // Add a mutex around to avoid the threads accessing the context simultaneously when resizing
+    
+    [self lockOpenGLContext];
+    
+    NSRect rect = [self convertRectToBacking:self.bounds];
+    
+    CCDirector *director = _director;
+    [director reshapeProjection: NSSizeToCGSize(rect.size) ];
+    
+    // avoid flicker
+    // Only draw if there is something to draw, otherwise it actually creates a flicker of the current glClearColor
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
+    
+    [self unlockOpenGLContext];
+}
+
+
+-(void) lockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    [glContext makeCurrentContext];
+    CGLLockContext([glContext CGLContextObj]);
+}
+
+-(void) unlockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    CGLUnlockContext([glContext CGLContextObj]);
+}
+
+// Find or make a fence that is ready to use.
+-(CCGLFence *)getReadyFence
+{
+    // First checkf oldest (first in the array) fence is ready again.
+    CCGLFence *fence = _fences.firstObject;;
+    if(fence.isReady){
+        // Remove the fence so it can be inserted at the end of the queue again.
+        [_fences removeObjectAtIndex:0];
+        return fence;
+    } else {
+        // No existing fences ready. Make a new one.
+        return [[CCGLFence alloc] init];
+    }
+}
+
+-(void)addFrameCompletionHandler:(dispatch_block_t)handler
+{
+    if(_fences == nil){
+        _fences = [NSMutableArray arrayWithObject:[[CCGLFence alloc] init]];
+    }
+    
+    CCGLFence *fence = _fences.lastObject;
+    if(!fence.isReady){
+        fence = [self getReadyFence];
+        [_fences addObject:fence];
+    }
+    
+    [fence.handlers addObject:handler];
+}
+
+-(void)beginFrame
+{
+    [self lockOpenGLContext];
+}
+
+-(void)presentFrame
+{
+    {
+        CCGLFence *fence = _fences.lastObject;
+        if(fence.isReady){
+            // If the fence is ready to be added, insert a sync point for it.
+            [fence insertFence];
+        }
+    }
+    
+    [self.openGLContext flushBuffer];
+    
+    // Check the fences for completion.
+    for(CCGLFence *fence in _fences){
+        if(fence.isCompleted){
+            for(dispatch_block_t handler in fence.handlers) handler();
+            [fence.handlers removeAllObjects];
+        } else {
+            break;
+        }
+    }
+    
+    [self unlockOpenGLContext];
+}
+
+-(GLuint)fbo
+{
+    return 0;
+}
+
+- (void) dealloc
+{
+    CCLOGINFO(@"cocos2d: deallocing %@", self);
+}
+
+#pragma mark CCGLView - Mouse Delegate
+
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```

Comment:
One of the biggest problems with the director being a singleton is that you can only have one cocos2D view in an application, and it makes the startup process *really* convoluted. (as I'm sure you've discovered) Since it's not accessible as a global anymore, it either needs to be passed around as an argument or set as a context value. Passing the director as an argument to all update methods, events, user callbacks, etc is also a little crazy and a *lot* more work for both us and users. Setting it as a context value lets it work in a more or less backwards compatible way for the majority of users that will only ever use a single view in their application.

We don't really have a lot of choice in forwarding the methods I don't think. Either every view type (currently 4 of them) needs to implement it's own event dispatching, or we need to forward the events to a common object that is shared by all of the platforms. Efficiency isn't really a problem. Events happen much less often than the rendering loop, so a few extra method calls aren't going to hurt the performance. It's Basically the same reason why we decided not to remove the director after thinking about it more. The code that is currently in the director would just need to be duplicated by 4 different view classes. On the other hand, we now have the opportunity to merge the 3 director classes into one since there isn't very much x-platform code in it anymore that couldn't move to the view classes.

Since we can't really break the dependence on having a director, or at the very least, knowing the view that a certain scene is running as a part of, we don't really have any other ideas of what to do.